File: two-number-sum/main.one.ts
// O(n) time | O(n) space

const twoNumberSum = (array: number[], targetSum: number) => {
	//create cache
	const hashTable: { [key: string]: boolean } = {};
	// iterate through array
	for (const v of array) {
		// unknown is target minus current iteree
		const unknown = targetSum - v;

		// if unknown is in hashtable, it exists earlier in array
		if (hashTable[unknown]) {
			return [unknown, v];
		} else {
			hashTable[v] = true;
		}
	}

	return [];
};

export {};

File: two-number-sum/main.one.cpp
#include <vector>
#include <unordered_set>
using namespace std;

vector<int> twoNumberSum(vector<int> array, int targetSum) {
  unordered_set<int> nums;
	
	for (int num = 0; num < array.size(); num++) {
		int unknown = targetSum - array[num];
		
		if (nums.find(unknown) != nums.end()) {
			return {unknown, array[num]};
		} else {
			nums.insert(array[num]);
		}
	}
  return {};
}
